src/models/library_by_type.rs

```diff
@@ -0,0 +1,189 @@
+                    .join(catalogs_effects)
+            }
+            Msg::Action(Action::Unload) => {
+                let selected_effects = eq_update(&mut self.selected, None);
+                let selectable_effects = selectable_update(&mut self.selectable, &self.selected);
+                let catalogs_effects = eq_update(&mut self.catalogs, vec![]);
+                selected_effects
+                    .join(selectable_effects)
+                    .join(catalogs_effects)
+            }
+            Msg::Action(Action::LibraryByType(ActionLibraryByType::LoadNextPage(index))) => {
+                match self.catalogs.get_mut(*index) {
+                    Some(catalog) => match (catalog.first(), catalog.last()) {
+                        (Some(first_page), Some(last_page))
+                            if !first_page.is_empty() && last_page.len() == CATALOG_PAGE_SIZE =>
+                        {
+                            let r#type = first_page
+                                .first()
+                                .map(|library_item| &library_item.r#type)
+                                .expect("first page of library catalog is empty");
+                            let skip = catalog.iter().fold(0, |result, page| result + page.len());
+                            let page = next_page::<F>(r#type, skip, &self.selected, &ctx.library);
+                            catalog.push(page);
+                            Effects::none()
+                        }
+                        _ => Effects::none().unchanged(),
+                    },
+                    _ => Effects::none().unchanged(),
+                }
+            }
+            Msg::Internal(Internal::LibraryChanged(_)) => {
+                catalogs_update::<F>(&mut self.catalogs, &self.selected, &ctx.library)
+            }
+            _ => Effects::none().unchanged(),
+        }
+    }
+}
+
+fn selectable_update(selectable: &mut Selectable, selected: &Option<Selected>) -> Effects {
+    let selectable_sorts = Sort::iter()
+        .map(|sort| SelectableSort {
+            sort: sort.to_owned(),
+            selected: selected
+                .as_ref()
+                .map(|selected| selected.sort == sort)
+                .unwrap_or_default(),
+        })
+        .collect();
+    let next_selectable = Selectable {
+        sorts: selectable_sorts,
+    };
+    eq_update(selectable, next_selectable)
+}
+
+fn catalogs_update<F: LibraryFilter>(
+    catalogs: &mut Vec<Catalog>,
+    selected: &Option<Selected>,
+    library: &LibraryBucket,
+) -> Effects {
+    let next_catalogs = match selected {
+        Some(selected) => {
+            let library_items = library
+                .items
+                .values()
+                .filter(|library_item| F::predicate(library_item))
+                .collect::<Vec<_>>();
+            library_items
+                .iter()
+                .map(|library_item| &library_item.r#type)
+                .unique()
```

Comment:
Is there something like `groupBy` in rust, to avoid iterating `library_items` n+1 times?